src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -434,20 +440,27 @@ void ZeroMQBackend::Run() {
                             break;
                         }
 
-                        // Empty result means xpub.send() returned EAGAIN. The
-                        // socket reached its high water mark and we should
-                        // relax / backoff a bit. Otherwise we'll be spinning
-                        // unproductively very fast here. Note that this is going
-                        // to build up backpressure and eventually inproc.send()
-                        // will block from the main thread.
+                        // Empty result means xpub.send() returned EAGAIN. The socket reached
+                        // its high-water-mark and we cannot send right now. We simply attempt
+                        // to re-send the message without the dontwait flag after increasing
+                        // the xpub stall metric. This way, ZeroMQ will block in xpub.send() until
+                        // there's enough room available.
                         if ( ! result ) {
-                            ++tries;
-                            auto sleep_for = std::min(tries * 10, 500);
-                            ZEROMQ_THREAD_PRINTF(
-                                "xpub: Failed forward inproc to xpub! Overloaded? (tries=%d sleeping %d ms)\n", tries,
-                                sleep_for);
-
-                            std::this_thread::sleep_for(std::chrono::milliseconds(sleep_for));
+                            total_xpub_stalls->Inc();
+
+                            try {
+                                zmq::send_flags block_flags =
+                                    zmq::send_flags::none | (flags & zmq::send_flags::sndmore);
+                                result = xpub.send(part, block_flags);
```

Comment:
Nice, I hadn't even thought about trying a non-blocking send first, then capture the failure to increase the metric and only then a blocking send, that's pretty clever. Could you add a comment here explaining the trick? E.g.,

```suggestion
                                // We sent non-blocking above so we are able to observe and report stalls in a metric.
                                // Now that we have done that switch to blocking send.
                                zmq::send_flags block_flags =
                                    zmq::send_flags::none | (flags & zmq::send_flags::sndmore);
                                result = xpub.send(part, block_flags);
```